junit-plugin/src/main/scala/org/scalajs/junit/plugin/ScalaJSJUnitPlugin.scala

```diff
@@ -0,0 +1,525 @@
+package org.scalajs.junit.plugin
+
+import scala.language.reflectiveCalls
+
+import scala.reflect.internal.Flags
+import scala.tools.nsc._
+import scala.tools.nsc.plugins.{
+  Plugin => NscPlugin, PluginComponent => NscPluginComponent
+}
+
+/* The Scala.js jUnit plugin is a way to overcome make class annotation
+ * information of any test class accessible to the Scala.js testing framework.
+ * This is the information that would usually be accessed through reflection.
+ *
+ *
+ * Test class:
+ *
+ * class Foo {
+ *   @Before def before(): Unit = {
+ *     // Initialize the instance before the tests
+ *   }
+ *   @Test def bar(): Unit = {
+ *     // assert some stuff
+ *   }
+ *   @Ignore("baz not implemented yet") @Test def baz(): Unit = {
+ *     // assert some other stuff
+ *   }
+ * }
+ * object Foo {
+ *   @BeforeClass def beforeClass(): Unit = {
+ *     // Initialize some global state for the tests.
+ *   }
+ * }
+ *
+ *
+ * Will be transformed to:
+ *
+ * class Foo extends org.scalajs.junit.ScalaJSJUnitTest {
+ *   @Before def before(): Unit = {
+ *     // Initialize the instance before the tests
+ *   }
+ *   @Test def bar(): Unit = {
+ *     // assert some stuff
+ *   }
+ *   @Ignore @Test def baz(): Unit = {
+ *     // assert some other stuff
+ *   }
+ *   def scalajs$junit$invoke(methodId: String): Unit = {
+ *     if (methodId = "0") before()
+ *     else if (methodId = "1") bar()
+ *     else if (methodId = "2") baz()
+ *     else throw new NoSuchMethodException(methodId)
+ *   }
+ * }
+ * object Foo extends org.scalajs.junit.ScalaJSJUnitTest {
+ *   @BeforeClass def beforeClass(): Unit = {
+ *     // Initialize some global state for the tests.
+ *   }
+ * }
+ * object Foo$scalajs$junit$hook extends org.scalajs.junit.JUnitTestMetadata {
+ *   def metadata(): JUnitClassMetadata = {
+ *     new JUnitClassMetadata(
+ *       annotations = List(),
+ *       moduleAnnotations = List(),
+ *       methods = List(
+ *           new JUnitMethodMetadata(name = "before", id = "0",
+ *               annotations = List(new Before)),
+ *           new JUnitMethodMetadata(name = "bar", id = "1",
+ *               annotations = List(new Test)),
+ *           new JUnitMethodMetadata(name = "baz", id = "2",
+ *               annotations = List(new Test, new Ignore("baz not implemented yet")))
+ *       ),
+ *       moduleMethods(
+ *           new JUnitMethodMetadata(name = "beforeClass", id = "0",
+ *               annotations = List(new BeforeClass)))
+ *     )
+ *   }
+ *   def newInstance(): ScalaJSJUnitTest = new Foo()
+ *   def scalajs$junit$invoke(methodId: String): Unit = {
+ *     if (methodId = "0") Foo.beforeClass()
+ *     else throw new NoSuchMethodException(methodId)
+ *   }
+ * }
+ *
+ * The test framework will identify Foo$scalajs$junit$hook as a test module
+ * because it extends JUnitTestMetadata. It will know which methods to run based
```

Comment:
The name `JUnitTestMetadata` seems to be misleading here. The whole module is not the metadata, it is the `hook`/`invoker`/`launcher`, whatever. What you call `JUnitClassMetadata` is the actual metadata. Consider renaming both of them.